File: functionTests/tyfxTests/tyxfTest5.R
library(TFunHDDC)
set.seed(1027)
#simulataed univariate data
data = genModelFD(ncurves=21, nsplines=21, alpha=c(0.9,0.9,0.9),
                  eta=c(10, 7, 17))
data
W=diag(rep(1,21))
W_m=diag(rep(1,21))
dety=1

Wlist<-list(W=W,
            W_m=W_m,
            dety=dety)
Wlist
K=3
Q<-vector(mode='list', length=3)
for (i in 1:K) Q[[i]]<-matrix(i*0.1,ncol=21, nrow=21)
Q
Q1<-vector(mode='list', length=3)
for (i in 1:K) Q1[[i]]<-matrix(i*0.1,ncol=21, nrow=21)
Q1
nsplines=21
mu=rbind(c(1, 0, 50, 100, rep(0, nsplines-4)),
         c(0, 0, 80, 0, 40, 2, rep(0, nsplines-6)),
         c(rep(0, nsplines-6), 20, 0, 80, 0, 0, 100))
mu
a=matrix(nrow=3,ncol=3,0)
a
a[1,]=c(6, 7, 1)
a[2,]=c(6.8, 3, 9.8)
a[3,]=c(1.2, 4, 2.5)
a
par=list(K=3,nux=c(201,201,201),a=a, b=c(4e-33,4.4e-65,0.1), d=c(3,3,3),mu=mu,prop=c(1/3,1/3,1/3),
         Q=Q,Q1=Q1)
res = TFunHDDC:::.T_funhddt_e_step1(data$fd,Wlist, par,clas=0,known=NULL, kno=NULL)

.T_tyxf7 <- function(dfconstr, nux, n, t, tw, K, p, N){
  if(dfconstr=="no"){
    dfoldg <- nux
    for(i in 1:K){
      constn <- 1 + (1/n[i]) * sum(t[, i] * ( log(tw[, i]) - tw[, i]) ) +
        digamma( (dfoldg[i] + p) / 2 ) - log( (dfoldg[i] + p) / 2 )
      temp = digamma( (dfoldg[i] + p) / 2 )
      print("digamma 7")
      print(temp)
      #print("7 no")
      #print(constn)
      nux[i] <- uniroot( function(v) log(v / 2) - digamma(v / 2) + constn,
                         lower = 0.0001, upper = 1000, tol = 0.00001)$root
      if(nux[i] > 200){
        nux[i] <- 200
      }
      if(nux[i] < 2){
        nux[i] <- 2
      }
    }
  }
  else
  {
    dfoldg <- nux[1]
    constn <- 1 + (1/N) * sum( t * (log(tw) - tw) ) +
      digamma( (dfoldg + p) / 2 ) - log( (dfoldg + p) / 2 )
    
    #print("7 yes")
    #print(constn)
    dfsamenewg <- uniroot( function(v) log(v / 2) - digamma(v / 2) + constn,
                           lower = 0.0001, upper = 1000, tol = 0.01)$root
    if(dfsamenewg > 200){
      dfsamenewg <- 200
    }
    if(dfsamenewg < 2){
      dfsamenewg <- 2
    }
    nux <- c( rep(dfsamenewg, K) )
  }
  return(nux)
} # end of .T_tyxf7

.T_tyxf8 <- function(dfconstr, nux, n, t, tw, K, p, N){
  if(dfconstr == "no"){
    dfoldg <- nux
    for(i in 1:K){
      constn <- 1 + (1 / n[i]) * sum( t[, i] * (log(tw[, i]) - tw[,i]) ) +
        digamma( (dfoldg[i] + p) / 2 ) -
        log( (dfoldg[i]+p) / 2 )
      temp1 = (dfoldg[i] + p) / 2 
      temp = digamma( (dfoldg[i] + p) / 2 )
      print('dfoldg')
      print(dfoldg)
      constn <- -constn
      print("digamma 8")
      print(temp)
      #print("8 no")
      #print(constn)
      nux[i] <- (-exp(constn) + 2 * ( exp(constn)) *
                   ( exp( digamma(dfoldg[i] / 2)) -
                       ( (dfoldg[i]/2) - (1/2) ) ) ) / ( 1 - exp(constn) )
      if(nux[i] > 200){
        nux[i] <- 200
      }
      if(nux[i] < 2){
        nux[i] <- 2
      }
    }
  } else {
    dfoldg <- nux[1]
    constn <- 1 + (1 / N) * sum( t * (log(tw) - tw) ) +
      digamma( (dfoldg + p) / 2 ) - log( (dfoldg + p) / 2 )
    constn <- -constn
    #print("8 yes")
    #print(constn)
    dfsamenewg <- (-exp(constn) + 2 * ( exp(constn)) *
                     (exp( digamma(dfoldg / 2)) -
                        ( (dfoldg /2) - (1 / 2) ) ) ) / ( 1 - exp(constn) )
    if(dfsamenewg > 200){
      dfsamenewg <- 200
    }
    if(dfsamenewg < 2){
      dfsamenewg <- 2
    }
    nux <- c( rep(dfsamenewg, K) )
  }
  return(nux)
} # end of .T_tyxf8

test17 = .T_tyxf7('no', par[['nux']], colSums(res[['t']]), res[['t']], res[['tw']], 3, 21, 21)
test27 = TFunHDDC:::.T_tyxf7('yes', par[['nux']], colSums(res[['t']]), res[['t']], res[['tw']], 3, 21, 21)

test18 = .T_tyxf8('no', par[['nux']], colSums(res[['t']]), res[['t']], res[['tw']], 3, 21, 21)
test28 = TFunHDDC:::.T_tyxf8('yes', par[['nux']], colSums(res[['t']]), res[['t']], res[['tw']], 3, 21, 21)

File: functionTests/tyfxTests/tyxfTest5.py
import numpy as np
import scipy.special as scip
import scipy.optimize as scio
import csv


def _T_tyxf7(dfconstr, nux, n, t, tw, K, p, N):
    newnux = np.zeros(3)
    if dfconstr == "no":
        dfoldg = nux.copy()

        #scipy digamma is slow? https://gist.github.com/timvieira/656d9c74ac5f82f596921aa20ecb6cc8
        for i in range(0, K):
            constn = 1 + (1/n[i]) * np.sum(t[:, i] * (np.log(tw[:, i]) - tw[:, i])) + scip.digamma((dfoldg[i] + p)/2) - np.log((dfoldg[i] + p)/2)
            temp = scip.digamma((dfoldg[i] + p)/2)
            print(f'digamma 7 value:{temp}')
            print(constn)
            f = lambda v : np.log(v/2) - scip.digamma(v/2) + constn
            #Verify this outputs the same as R: may need to set rtol to 0
            newnux[i] = scio.brentq(f, 0.0001, 1000, xtol=0.00001)

            if newnux[i] > 200:
                newnux[i] = 200.

            if newnux[i] < 2:
                newnux[i] = 2.

    else:
        dfoldg = nux[0]
        constn = 1 + (1/N) * np.sum(t *(np.log(tw) - tw)) + scip.digamma( (dfoldg + p) / 2) - np.log( (dfoldg + p) / 2)

        print(constn)
        f = lambda v : np.log(v/2) - scip.digamma(v/2) + constn
            #Verify this outputs the same as R: may need to set rtol to 0
        dfsamenewg = scio.brentq(f, 0.0001, 1000, xtol=0.01)

        if dfsamenewg > 200:
            dfsamenewg = 200.
        
        if dfsamenewg < 2:
            dfsamenewg = 2.

        newnux = np.repeat(dfsamenewg, K)

    return newnux

def _T_tyxf8(dfconstr, nux, n, t, tw, K, p, N):
    newnux = np.zeros(3)

    if(dfconstr == "no"):
        dfoldg = nux.copy()
        
        for i in range(0, K):
            constn = 1 + (1 / n[i]) * np.sum(t[:, i] * (np.log(tw[:, i]) - tw[:, i])) + scip.digamma((dfoldg[i] + p)/2) - np.log( (dfoldg[i] + p)/2)
            temp = scip.digamma((dfoldg[i] + p)/2)
            print('dfoldg')
            print(dfoldg)
            #print((dfoldg[i]+p)/2)
            #print(f'digamma 8 value:{temp}')
            constn = -constn
            print(constn)
            newnux[i] = (-np.exp(constn) + 2 * (np.exp(constn)) * (np.exp(scip.digamma(dfoldg[i] / 2)) - ( (dfoldg[i]/2) - (1/2)))) / (1 - np.exp(constn))

            if newnux[i] > 200:
                newnux[i] = 200.

            if newnux[i] < 2:
                newnux[i] = 2.

    else:
        dfoldg = nux[0]
        constn = 1 + (1 / N) * np.sum(t * (np.log(tw) - tw)) + scip.digamma((dfoldg + p)/2) - np.log( (dfoldg + p)/2)
        constn = -constn

        print(constn)
        dfsamenewg = (-np.exp(constn) + 2 * (np.exp(constn)) * (np.exp(scip.digamma(dfoldg / 2)) - ( (dfoldg/2) - (1/2)))) / (1 - np.exp(constn))

        if dfsamenewg > 200:
            dfsamenewg = 200.

        if dfsamenewg < 2:
            dfsamenewg = 2.

        newnux = np.repeat(dfsamenewg, K)

    return newnux

data = []
mats = {'t': [], 'tw': []}
with open('res5.csv', newline='') as csvfile:
    datareader = csv.reader(csvfile, delimiter=' ', quotechar='|')
    for row in datareader:
        data.append(row[0].split(',')[1:])

#print(data)
#print(data[1][0:3])
for i in data[1:]:
    mats['t'].append(i[0:3])
    mats['tw'].append(i[3:6])

mats['L'] = np.array(data[1][-1]).astype(float)

mats['t'] = np.array(mats['t']).astype(float)
mats['tw'] = np.array(mats['tw']).astype(float)

par = {'K': 3, 'nux':np.array([201.,201.,201.])}


test1_7 = _T_tyxf7('no', par['nux'], np.sum(mats['t'], axis=0), mats['t'], mats['tw'], 3, 21., 21.)
test2_7 = _T_tyxf7('yes', par['nux'], np.sum(mats['t'], axis=0), mats['t'], mats['tw'], 3, 21., 21.)

test1_8 = _T_tyxf8('no', par['nux'], np.sum(mats['t'], axis=0), mats['t'], mats['tw'], 3, 21., 21.)
test2_8 = _T_tyxf8('yes', par['nux'], np.sum(mats['t'], axis=0), mats['t'], mats['tw'], 3, 21., 21.)


print(f'tyfx7,dfconstr=no:{test1_7}; tyfx7,dfconstr=yes:{test2_7};tyfx8,dfconstr=no:{test1_8};tyfx8,dfconstr=yes:{test2_8}')